src/Statement/Grouper/FilteringStatementGrouper.php

```diff
@@ -0,0 +1,101 @@
+<?php
+
+namespace Wikibase\DataModel\Services\Statement\Grouper;
+
+use InvalidArgumentException;
+use Wikibase\DataModel\Statement\Statement;
+use Wikibase\DataModel\Statement\StatementFilter;
+use Wikibase\DataModel\Statement\StatementList;
+
+/**
+ * @since 3.2
+ *
+ * @licence GNU GPL v2+
+ * @author Thiemo Mättig
+ */
+class FilteringStatementGrouper implements StatementGrouper {
+
+	/**
+	 * @var StatementList[]
+	 */
+	private $groups = array();
```

Comment:
I'm not happy with this too, because it means statements will stack up when calling this filter multiple times. Since I'm quite proud of the tests for this class, feel free to refactor it if you can do this without touching the tests.